File: src/solar/solar-common/util/debug/DebugVmUtil.ts
/** @noSelfInFile */
import FunctionInfo = debug.FunctionInfo;

export default class DebugVmUtil {

    /**
     * * 'c': the hook is called every time Lua calls a function;
     * * 'r': the hook is called every time Lua returns from a function;
     * * 'l': the hook is called every time Lua enters a new line of code.
     * 'call' | 'return' | 'line' | 'count'
     */
    static sethook(hook: (event: 'call' | 'return' | 'line' | 'count', line?: number) => any, mask: string) {
        debug.sethook(hook, mask, 0)
    }


    /**
     * 根据函数调用对战 获得相关的用户脚本文件加载 路径
     */
    static getUserScriptRequireModNameByStack(): string[] {
        let userScriptPaths: string[] = []
        for (let i = 1; i < 200; i++) {
            let requireModName = DebugVmUtil.getRequireModName(i);
            if (requireModName == null) {
                return userScriptPaths;
            }
            //
            if (requireModName == "") {
                continue;
            }
            if (requireModName.includes("solar.")) {
                continue;
            }
            if (requireModName.includes("lualib_bundle")) {
                continue;
            }
            if (requireModName.includes("App")) {
                continue;
            }
            //
            if (!userScriptPaths.includes(requireModName)) {
                userScriptPaths.push(requireModName)
            }

        }
        return userScriptPaths;
    }


    /**
     * 获取当前执行文件的 require 的路径
     * @param f
     */
    static getRequireModName(f: number): string {
        let information = debug.getinfo(f, "nS");
        if (information == null) {
            return null;
        }
        let luaPath = information.source.substring(1, information.source.length - 4);
        luaPath = luaPath.replaceAll(local_map_dir_path, "");
        luaPath = luaPath.replaceAll("scripts\\", "");
        return luaPath.replaceAll("\\", ".");
    }

    /**
     * 获取函数信息对应的TS的代码行
     * @param functionInfo
     */
    static getTsLine(functionInfo: FunctionInfo) {
        if (functionInfo.short_src == null) {
            return functionInfo.linedefined;
        }
        let tsLine = tostring(functionInfo.linedefined || 0)
        if (_G['__TS__sourcemap']) {
            if (_G['__TS__sourcemap'][functionInfo.short_src]) {
                tsLine = tostring(_G['__TS__sourcemap'][functionInfo.short_src][tsLine] || tsLine)
            }
        }
        return tsLine;
    }

    /**
     * 获取TS代码文件路径与名字与代码行
     * @param functionInfo
     */
    static getTsSrcLineName(functionInfo: FunctionInfo) {
        if (!functionInfo) {
            return "";
        }
        let src = functionInfo.source;
        if (!src) {
            src = "<C>";
        } else if (src.substring(src.length - 4, src.length) == ".lua") {
            src = src.substring(0, src.length - 4);
        }
        let name = functionInfo.name;
        if (!name) {
            name = "匿名";
        } else if (name.substring(name.length - 2, name.length) == "_l") {
            name = name.substring(0, name.length - 2);
        }
        let title = src + ":" + tostring(DebugVmUtil.getTsLine(functionInfo)) + ":" + name;
        return title;
    }


    /**
     * 勾住函数调用与返回 并打印信息
     * (用作崩溃查询 看崩溃时调用到哪个函数了)
     * @param writeFile 将最近的函数调用写到文件里
     * @param printMaxCount 每秒内相同函数最大打印次数
     * @param mask c/cr c=函数开始调用时   r=函数结束返回时
     */
    static hookCRAndPrintInfo(writeFile: boolean = true, printMaxCount: number = 2, mask: string = "c") {
        let pt: { [name: string]: number } = {}
        let [file] = io.open("_sre/temp/hookCRAndPrintInfo.txt", "wb");
        let cLines = []
        DebugVmUtil.sethook((e) => {
            let information = debug.getinfo(2, "nS");
            let tsSrcNameLine = DebugVmUtil.getTsSrcLineName(information as any);
            let count = (pt[tsSrcNameLine] || 0) + 1
            pt[tsSrcNameLine] = count
            if (printMaxCount < 0 || count < printMaxCount) {
                print(tsSrcNameLine)
            }
            if (time % 1000 == 0) {
                pt = {};
            }
            if (!writeFile) {
                return;
            }
            if (cLines.length > 1000) {
                cLines = cLines.slice(cLines.length - 20, cLines.length)
            }
            cLines.push(tsSrcNameLine)
            file.seek("set")
            file.write("\uFEFF")
            for (let i = Math.max(cLines.length - 20, 0); i < cLines.length; i++) {
                file.write(cLines[i] + "\r\n")
            }
            file.write("\r\n####结束####\r\n")
            file.flush();
        }, mask)
    }
}

File: scripts/solar/solar-common/util/debug/DebugVmUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__StringIncludes = ____lualib.__TS__StringIncludes
local __TS__ArrayIncludes = ____lualib.__TS__ArrayIncludes
local __TS__StringSubstring = ____lualib.__TS__StringSubstring
local __TS__StringReplaceAll = ____lualib.__TS__StringReplaceAll
local __TS__ArraySlice = ____lualib.__TS__ArraySlice
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["11"] = 4,["12"] = 4,["13"] = 4,["15"] = 4,["16"] = 12,["17"] = 13,["18"] = 12,["19"] = 20,["20"] = 21,["22"] = 22,["23"] = 22,["25"] = 23,["26"] = 24,["27"] = 25,["29"] = 28,["30"] = 29,["32"] = 31,["33"] = 32,["35"] = 34,["36"] = 35,["38"] = 37,["39"] = 38,["41"] = 41,["42"] = 42,["46"] = 22,["49"] = 46,["50"] = 20,["51"] = 54,["52"] = 55,["53"] = 56,["54"] = 57,["56"] = 59,["57"] = 60,["58"] = 61,["59"] = 62,["60"] = 54,["61"] = 69,["62"] = 70,["63"] = 71,["65"] = 73,["66"] = 74,["67"] = 75,["68"] = 76,["71"] = 79,["72"] = 69,["73"] = 86,["74"] = 87,["75"] = 88,["77"] = 90,["78"] = 91,["79"] = 92,["80"] = 93,["81"] = 94,["83"] = 96,["84"] = 97,["85"] = 98,["86"] = 99,["87"] = 100,["89"] = 102,["90"] = 103,["91"] = 86,["92"] = 114,["93"] = 114,["94"] = 114,["96"] = 114,["97"] = 114,["99"] = 114,["100"] = 114,["102"] = 115,["103"] = 116,["104"] = 117,["105"] = 118,["106"] = 118,["107"] = 119,["108"] = 120,["109"] = 121,["110"] = 122,["111"] = 123,["112"] = 124,["114"] = 126,["115"] = 127,["117"] = 129,["120"] = 132,["121"] = 133,["123"] = 135,["124"] = 136,["125"] = 137,["127"] = 138,["128"] = 138,["129"] = 139,["130"] = 138,["133"] = 141,["134"] = 142,["135"] = 118,["136"] = 118,["137"] = 118,["138"] = 114});
local ____exports = {}
____exports.default = __TS__Class()
local DebugVmUtil = ____exports.default
DebugVmUtil.name = "DebugVmUtil"
function DebugVmUtil.prototype.____constructor(self)
end
function DebugVmUtil.sethook(self, hook, mask)
    debug.sethook(hook, mask, 0)
end
function DebugVmUtil.getUserScriptRequireModNameByStack(self)
    local userScriptPaths = {}
    do
        local i = 1
        while i < 200 do
            do
                local requireModName = ____exports.default:getRequireModName(i)
                if requireModName == nil then
                    return userScriptPaths
                end
                if requireModName == "" then
                    goto __continue5
                end
                if __TS__StringIncludes(requireModName, "solar.") then
                    goto __continue5
                end
                if __TS__StringIncludes(requireModName, "lualib_bundle") then
                    goto __continue5
                end
                if __TS__StringIncludes(requireModName, "App") then
                    goto __continue5
                end
                if not __TS__ArrayIncludes(userScriptPaths, requireModName) then
                    userScriptPaths[#userScriptPaths + 1] = requireModName
                end
            end
            ::__continue5::
            i = i + 1
        end
    end
    return userScriptPaths
end
function DebugVmUtil.getRequireModName(self, f)
    local information = debug.getinfo(f, "nS")
    if information == nil then
        return nil
    end
    local luaPath = __TS__StringSubstring(information.source, 1, #information.source - 4)
    luaPath = __TS__StringReplaceAll(luaPath, local_map_dir_path, "")
    luaPath = __TS__StringReplaceAll(luaPath, "scripts\\", "")
    return __TS__StringReplaceAll(luaPath, "\\", ".")
end
function DebugVmUtil.getTsLine(self, functionInfo)
    if functionInfo.short_src == nil then
        return functionInfo.linedefined
    end
    local tsLine = tostring(functionInfo.linedefined or 0)
    if _G.__TS__sourcemap then
        if _G.__TS__sourcemap[functionInfo.short_src] then
            tsLine = tostring(_G.__TS__sourcemap[functionInfo.short_src][tsLine] or tsLine)
        end
    end
    return tsLine
end
function DebugVmUtil.getTsSrcLineName(self, functionInfo)
    if not functionInfo then
        return ""
    end
    local src = functionInfo.source
    if not src then
        src = "<C>"
    elseif __TS__StringSubstring(src, #src - 4, #src) == ".lua" then
        src = __TS__StringSubstring(src, 0, #src - 4)
    end
    local name = functionInfo.name
    if not name then
        name = "匿名"
    elseif __TS__StringSubstring(name, #name - 2, #name) == "_l" then
        name = __TS__StringSubstring(name, 0, #name - 2)
    end
    local title = (((src .. ":") .. tostring(____exports.default:getTsLine(functionInfo))) .. ":") .. name
    return title
end
function DebugVmUtil.hookCRAndPrintInfo(self, writeFile, printMaxCount, mask)
    if writeFile == nil then
        writeFile = true
    end
    if printMaxCount == nil then
        printMaxCount = 2
    end
    if mask == nil then
        mask = "c"
    end
    local pt = {}
    local file = io.open("_sre/temp/hookCRAndPrintInfo.txt", "wb")
    local cLines = {}
    ____exports.default:sethook(
        function(e)
            local information = debug.getinfo(2, "nS")
            local tsSrcNameLine = ____exports.default:getTsSrcLineName(information)
            local count = (pt[tsSrcNameLine] or 0) + 1
            pt[tsSrcNameLine] = count
            if printMaxCount < 0 or count < printMaxCount then
                print(tsSrcNameLine)
            end
            if time % 1000 == 0 then
                pt = {}
            end
            if not writeFile then
                return
            end
            if #cLines > 1000 then
                cLines = __TS__ArraySlice(cLines, #cLines - 20, #cLines)
            end
            cLines[#cLines + 1] = tsSrcNameLine
            file:seek("set")
            file:write("﻿")
            do
                local i = math.max(#cLines - 20, 0)
                while i < #cLines do
                    file:write(tostring(cLines[i + 1]) .. "\r\n")
                    i = i + 1
                end
            end
            file:write("\r\n####结束####\r\n")
            file:flush()
        end,
        mask
    )
end
return ____exports
